### PR TITLE
Add an option to exclude options based on regex.

### PR DESCRIPTION
**Pull request by timuralp**
Adds an "exclude_pattern" option for sync/lifecycle that allows
skipping objects that match the specified regular expression. This is
useful in case objects with certain names have to be skipped (for
example, if they are segments of a large object).

### CODE REVIEW

File: test/integration/test_s3_sync.py [hunk 1]
```diff
@@ -1039,6 +1041,49 @@ def test_swift_sync_same_segments_slo(self):
...
+            'get_object', mapping['aws_bucket'], key)
+        self.assertEqual(body, content)
+        self.assertEqual(
+            '"%s"' % hashlib.md5('%s%s' % (
+                hashlib.md5(content[:1024]).hexdigest(),
+                hashlib.md5(content[1024:]).hexdigest())).hexdigest(),
+            headers['etag'])
+
+        _, listing = self.remote_swift('get_container', mapping['aws_bucket'])
+        self.assertEqual(1, len(listing))
```
Comment (thiagodasilva) [inline test/integration/test_s3_sync.py]: just for sanity, we should also add a `get_container` of the segments container...

File: s3_sync/sync_container.py [hunk 2]
```diff
@@ -192,6 +202,11 @@ def handle_container_metadata(self, metadata_dict, db_id):
                 ''.join(traceback.format_tb(post_resp.exc_info[2])))
 
     def handle(self, row, swift_client):
+        if self.exclude_regex.match(row['name']) is not None:
+            self.logger.debug('Skipping excluded object: %s/%s' % (
+                self._container, row['name'].decode('utf-8')))
```
[the remark this answers is not in the document]
Comment (timuralp) [reply s3_sync/sync_container.py]: Are you sure? It comes from the JSON config and should be a unicode string. E.g:
```
>>> json.loads('{"container": "fo\xc3\xa4"}')
{u'container': u'fo\xe4'}
```

File: test/integration/test_s3_sync.py (hunk 1 above)
Comment (timuralp) [reply test/integration/test_s3_sync.py]: What are we trying to ascertain? I feel like checking the body verifies the content already and we have other tests to verify the actual DLO transfers?

File: docs/source/running.rst [hunk 3]
```diff
@@ -75,6 +76,11 @@ Sync Profile
   - **protocol**: remote object store API protocol: ``swift`` or ``s3``.
   - **copy_after**: Time in seconds to delay object sync (*Optional*.
     Default: 0).
+  - **exclude_pattern**: Regular expression to be applied to object names to
```
[the remark this answers is not in the document]
Comment (timuralp) [reply docs/source/running.rst]: I think that's probably a separate document that explains the caveats and limitations. I don't think we need to add it in the option description itself.

[the remark this answers is not in the document]
Comment (thiagodasilva) [reply docs/source/running.rst]: ok

File: s3_sync/sync_container.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thiagodasilva) [reply s3_sync/sync_container.py]: my mistake, I thought I ran into this problem with another patch.

File: test/integration/test_s3_sync.py (hunk 1 above)
Comment (thiagodasilva) [reply test/integration/test_s3_sync.py]: good point, missed the body verification.

**Pull request merged**